### PR TITLE
JS: Join-order improvements preparing for new SmartInliner

### PR DESCRIPTION
**Pull request by erik-krogh**
Unblocks the new inliner.   

### old cli + old QL vs. new cli + new QL:  
Should confirm that things are good when the new CLI goes live.   
I ran [default/TaintedPath.ql](https://github.com/dsp-testing/erik-krogh-dca/tree/run/inliner-5/reports), [nightly/security](https://github.com/dsp-testing/erik-krogh-dca/tree/run/inliner-6/reports), and [default/lgtm-full](https://github.com/dsp-testing/erik-krogh-dca/tree/run/inliner-esbens-suggested-run/reports).  
All look fine. 

### old cli + old QL vs. old cli + new QL:  

Should confirm that the new QL will also perform well with the old CLI.   
[nightly/security](https://github.com/dsp-testing/erik-krogh-dca/tree/run/inliner-7/reports) looks fine.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll [hunk 1]
```diff
@@ -96,6 +96,7 @@ module DataFlow {
     predicate accessesGlobal(string g) { globalVarRef(g).flowsTo(this) }
 
     /** Holds if this node may evaluate to the string `s`, possibly through local data flow. */
+    pragma[nomagic]
     predicate mayHaveStringValue(string s) {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll:100]: my gut feeling is that we should run lgtm-full.qls/default.yml for this change. Or am I just being paranoid?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll:100]: Lets do it! https://github.com/dsp-testing/erik-krogh-dca/tree/run/inliner-esbens-suggested-run/reports  

There will definitely be some 6 hour timeouts in that run.

Comment (esbena) [reply javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll:100]: wow. I am quite happy to be so wrong!

**Comment by ginsbach** [comment]
I can confirm that this has fixed the problems.:
```
Sat 10 Apr 00:34:23 BST 2021
Using default QL search-path.
baseline is 0ccc86b846ba8dbc6bb13fd122a9b9a527608387 Merge pull request #39072 from github/missing-eclipse-project-dependency
branch1 is 103f5dcc413b60c2621b1fa328b808731c5d61c4 Merge 8e1cd5d6d8ea8488c922c20fd07b86d4a59ae9f8 into 0ccc86b846ba8dbc6bb13fd122a9b9a527608387
Results for branch1 compare OK.

Runtime                                         baseline          branch1        
---------------------------------------------------------------------------------
cpp-apache-codingstars                            395183    408432  +3.35% +0.038
cpp-gcc-codingstars                               345400    361759  +4.74% +0.013
cpp-git-codingstars                               500907    503916  +0.60% +0.316
csharp-Semmle_ql_1b25573                          391563    378889  -3.24% -0.039
csharp-dotnet_roslyn_b4a4a98                     3034064   2683205 -11.56% -0.001
csharp-mono_monodevelop_cb6407d                  4251410   3410709 -19.77% -5e-05
java-hadoop-codingstars                          1437471   1360419  -5.36% -0.009
java-vavr-revision-2017-September-13--22-30-24    476492    486375  +2.07% +0.102
javascript-ChakraCore-revision-d9d540d           2979469   3030180  +1.70% +0.136
javascript-bwip-js-revision-d9d540d               439178    436386  -0.64% -0.307
javascript-rhino-revision-d9d540d                 629329    621397  -1.26% -0.188
python-kivy-codingstars                           361488    368199  +1.86% +0.121
python-letsencrypt-codingstars                    165441    161655  -2.29% -0.083
python-openstack-cinder-codingstars              1832590   1777172  -3.02% -0.046
---------------------------------------------------------------------------------
TOTAL:                                          17239985  15988693  -7.26% -4e-05

```

**Review by esbena — APPROVED** [review]

**Pull request merged**